ttnn/cpp/ttnn/operations/data_movement/untilize/device/untilize_program_factory.cpp

```diff
@@ -34,13 +34,36 @@ operation::ProgramWithCallbacks untilize_multi_core(
     Device* device = a.device();
 
     uint32_t ntiles = a.volume() / TILE_HW;
-    uint32_t ntiles_per_block = a.get_legacy_shape()[-1] / TILE_WIDTH;
-    uint32_t nblocks = ceil((float)ntiles / ntiles_per_block);
-    uint32_t block_size_nbytes = a.get_legacy_shape()[-1] * output.element_size();
+    uint32_t stick_s = a.get_legacy_shape()[-1];
+    uint32_t stick_size = stick_s * output.element_size();
+    uint32_t ntiles_in_row = stick_s / TILE_WIDTH;
+    uint32_t max_l1_size = a.device()->l1_size_per_core() / 2 - L1_UNRESERVED_BASE;
+    uint32_t max_tiles = (max_l1_size / (input_single_tile_size + output_single_tile_size))/2;  // 2 CBs, double buffering each
+    // Currently need the number of tiles in a row to be divisible by tiles in a block
+    uint32_t ntiles_per_block = 1;
+    if (ntiles_in_row <= max_tiles) {
+        ntiles_per_block = ntiles_in_row;
+    } else {
+        for (uint32_t n_t = max_tiles; n_t > 0; n_t--) {
+            if (ntiles_in_row % n_t == 0) {
+                ntiles_per_block = n_t;
+                break;
+            }
+        }
+    }
```

Comment:
you can simplify to:
```suggestion
    uint32_t ntiles_per_block = ntiles_in_row;
    if (ntiles_in_row > max_tiles) {
        for (uint32_t n_t = max_tiles; n_t > 0; n_t--) {
            if (ntiles_in_row % n_t == 0) {
                ntiles_per_block = n_t;
                break;
            }
        }
    }
```